File: src/main/kotlin/fit/wenchao/db/JavaPackage.java
package fit.wenchao.db;

import java.io.File;
import java.io.IOException;
import java.io.OutputStream;
import java.nio.file.Files;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

public class JavaPackage {
    private String absolutePackageString;

    private String relativePackageString;

    public static JavaPackage from(String relativePackageString) {
        JavaPackage javaPackage = new JavaPackage();

        File file = new File("");
        String projectPath = file.getAbsolutePath();

        javaPackage.relativePackageString = relativePackageString;
        javaPackage.setPackage(projectPath + "/src/main/kotlin/" + relativePackageString);
        if(!new File(javaPackage.absolutePackageString).exists()) {
            try {
                Files.createDirectories(new File(javaPackage.absolutePackageString).toPath());
            }
            catch (IOException e) {
                throw new RuntimeException(e);
            }
        }
        return javaPackage;
    }

    public static void main(String[] args) {
        System.out.println(JavaPackage.from ("fit/wenchao/db"));
    }

    private void setPackage(String packageString) {
        this.absolutePackageString = packageString;
    }

    @Override
    public String toString() {
        return "JavaPackage{" +
                "packageString='" + absolutePackageString + '\'' +
                '}';
    }

    public OutputStream getOutputStream() throws IOException {

        File file = new File(this.absolutePackageString);
        if(file.exists()) {
            return Files.newOutputStream(file.toPath());
        }
        throw new RuntimeException("package not exists");
    }

    public List<String> listSourceFileNames() {
        File file = new File(absolutePackageString);
        String[] list = file.list();
        List<String> sourceFileNames = new ArrayList<>();
        if(list != null) {
            sourceFileNames = Arrays.stream(list).collect(Collectors.toList());
        }
        return sourceFileNames;
    }

    public File getFile() {
        return new File(this.absolutePackageString);
    }

    public String getDotSplitName() {
        return relativePackageString.replace("/", ".");
    }
}

File: src/main/kotlin/fit/wenchao/db/JavaPackage.kt
package fit.wenchao.db

import com.fasterxml.jackson.databind.RuntimeJsonMappingException
import java.io.File
import java.io.IOException
import java.nio.file.Files
import java.util.*
import java.util.stream.Collectors

class JavaPackage {

    companion object {

        lateinit var projectToSrc: String

        init {
            // read projectToSrc from config.properties
            val props = Properties()
            val inputStream = JavaPackage::class.java.classLoader.getResourceAsStream("config.properties")
            inputStream?.let {
                props.load(inputStream)
            } ?: run {
                throw RuntimeException("config.properties not found")
            }
            val projectToSrc: String? = props.getProperty("projectToSrc")
            projectToSrc?.let {
                this.projectToSrc = projectToSrc
            } ?: run {
                throw RuntimeException("projectToSrc not found in config.properties")
            }
        }

        fun from(relativePackageString: String): JavaPackage {
            val javaPackage = JavaPackage()
            // get project root file
            val currentProjectRootFile = File("")
            val projectPath = currentProjectRootFile.absolutePath
            javaPackage.relativePackageString = relativePackageString
            javaPackage.absolutePackageString = "$projectPath/${projectToSrc}/$relativePackageString"
            if (!File(javaPackage.absolutePackageString).exists()) {
                try {
                    Files.createDirectories(File(javaPackage.absolutePackageString).toPath())
                } catch (e: IOException) {
                    throw RuntimeException(e)
                }
            }
            return javaPackage
        }
    }


    var absolutePackageString: String? = null
    var relativePackageString: String? = null


    override fun toString(): String {
        return "JavaPackage{" +
                "packageString='" + absolutePackageString + '\'' +
                '}'
    }

    fun listSourceFileNames(): List<String?> {
        val file = File(absolutePackageString)
        val list = file.list()
        var sourceFileNames: List<String?> = ArrayList()
        if (list != null) {
            sourceFileNames = Arrays.stream(list).collect(Collectors.toList())
        }
        return sourceFileNames
    }

    val file: File
        get() = File(absolutePackageString)

    val dotSplitName: String
        get() = relativePackageString!!.replace("/", ".")


}

